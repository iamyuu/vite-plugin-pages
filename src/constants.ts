export const MODULE_IDS = [
  '~react-pages',
  'virtual:generated-pages-react',
]

export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages'
export const ROUTE_IMPORT_NAME = '__pages_import_$1__'

export const dynamicRouteRE = /^\[(.+)\]$/
export const cacheAllRouteRE = /^\[\.{3}(.*)\]$/
export const replaceDynamicRouteRE = /^\[(?:\.{3})?(.*)\]$/

export const nuxtDynamicRouteRE = /^_(.*)$/
export const nuxtCacheAllRouteRE = /^_$/

export const countSlashRE = /\//g

export const replaceIndexRE = /\/?index$/
