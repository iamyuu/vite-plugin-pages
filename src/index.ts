import { MODULE_ID_VIRTUAL } from './constants'
import { PageContext } from './context'

import { parsePageRequest } from './utils'
import type { UserOptions } from './types'
import type { Plugin } from 'vite'

function pagesPlugin(userOptions: UserOptions = {}): Plugin {
  let ctx: PageContext

  return {
    name: 'vite-plugin-pages',
    enforce: 'pre',
    async configResolved(config) {
      userOptions.resolver = 'react'

      ctx = new PageContext(userOptions, config.root)
      ctx.setLogger(config.logger)
      await ctx.searchGlob()
    },
    configureServer(server) {
      ctx.setupViteServer(server)
    },
    resolveId(id) {
      if (ctx.options.moduleIds.includes(id))
        return `${MODULE_ID_VIRTUAL}?id=${id}`

      return null
    },
    async load(id) {
      const {
        moduleId,
        pageId,
      } = parsePageRequest(id)

      if (moduleId === MODULE_ID_VIRTUAL && pageId && ctx.options.moduleIds.includes(pageId))
        return ctx.resolveRoutes()

      return null
    },
  }
}

export * from './types'
export type { ReactRoute } from './resolvers'

export { syncIndexResolver } from './options'
export { PageContext }
export default pagesPlugin
