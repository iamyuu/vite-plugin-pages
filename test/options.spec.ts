import { describe, expect, test } from 'vitest'
import { resolveOptions } from '../src/options'

describe('Options resolve', () => {
  test('react', () => {
    const options = resolveOptions({
      dirs: 'examples/next-style/src/pages',
      resolver: 'react',
    })
    expect(options).toMatchSnapshot({
      root: expect.any(String),
    })
  })
})
