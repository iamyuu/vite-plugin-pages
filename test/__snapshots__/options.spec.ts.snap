// Vitest Snapshot v1

exports[`Options resolve > react 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/next-style/src/pages",
    },
  ],
  "exclude": [],
  "extendRoute": undefined,
  "extensions": [
    "tsx",
    "jsx",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(tsx\\|jsx\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "moduleIds": [
    "~react-pages",
    "virtual:generated-pages-react",
  ],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": {
    "resolveExtensions": [Function],
    "resolveModuleIds": [Function],
    "resolveRoutes": [Function],
    "stringify": {
      "component": [Function],
      "dynamicImport": [Function],
      "final": [Function],
    },
  },
  "root": Any<String>,
  "routeStyle": "next",
}
`;
